tool/lib/model.dart

```diff
@@ -0,0 +1,161 @@
+// Copyright 2020 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:io';
+
+import 'package:path/path.dart' as path;
+
+class DevToolsRepo {
+  DevToolsRepo._create(this.repoPath);
+
+  final String repoPath;
+
+  @override
+  String toString() => '[DevTools $repoPath]';
+
+  /// This returns the DevToolsRepo instance based on the current working
+  /// directory.
+  ///
+  /// This can fail and return null if the current working directory is not
+  /// contained within a git checkout of DevTools.
+  static DevToolsRepo getInstance() {
+    final String repoPath = _findRepoRoot(Directory.current);
+    return repoPath == null ? null : DevToolsRepo._create(repoPath);
+  }
+
+  List<Package> getPackages() {
+    final List<Package> result = [];
+
+    _collectPackages(Directory(repoPath), result);
+    result.sort((a, b) => a.packagePath.compareTo(b.packagePath));
+
+    return result;
+  }
+
+  static String _findRepoRoot(Directory dir) {
+    // Look for README.md, packages, tool.
+    if (_fileExists(dir, 'README.md') &&
+        _dirExists(dir, 'packages') &&
+        _dirExists(dir, 'tool')) {
+      return dir.path;
+    }
+
+    if (dir.path == dir.parent.path) {
+      return null;
+    } else {
+      return _findRepoRoot(dir.parent);
+    }
+  }
+
+  void _collectPackages(Directory dir, List<Package> result) {
+    if (_fileExists(dir, 'pubspec.yaml')) {
+      result.add(Package._(this, dir.path));
+    }
+
+    for (FileSystemEntity entity
+        in dir.listSync(recursive: false, followLinks: false)) {
+      final String name = path.basename(entity.path);
+      if (entity is Directory && !name.startsWith('.') && name != 'build') {
+        _collectPackages(entity, result);
+      }
+    }
+  }
+}
+
+class FlutterSdk {
+  FlutterSdk._(this.sdkPath);
+
+  /// Return the Flutter SDK.
+  ///
+  /// This can return null if the Flutter SDK can't be found.
+  static FlutterSdk getSdk() {
+    // Look for it relative to the current Dart process.
+    final String dartVmPath = Platform.resolvedExecutable;
+    final List<String> pathSegments = path.split(dartVmPath);
+    final List<String> expectedSegments =
+        path.posix.split('bin/cache/dart-sdk/bin/dart');
+
+    if (pathSegments.length >= expectedSegments.length) {
+      // Remove the trailing 'dart'.
+      pathSegments.removeLast();
+      expectedSegments.removeLast();
```

Comment:
why is it necessary for this to be done outside of the while loop

Reply:
This remove the last segment of the path - `Platform.resolvedExecutable` returns the path to the Dart VM; this is removing the name of that binary.